openstack/utils/choose_version.go

```diff
@@ -109,3 +113,123 @@ func ChooseVersion(client *gophercloud.ProviderClient, recognized []*Version) (*
 
 	return highest, endpoint, nil
 }
+
+type SupportedMicroversions struct {
+	MaxMajor int
+	MaxMinor int
+	MinMajor int
+	MinMinor int
+}
+
+// GetSupportedMicroversions returns the minimum and maximum microversion that is supported by the ServiceClient Endpoint.
+func GetSupportedMicroversions(client *gophercloud.ServiceClient) (SupportedMicroversions, error) {
+	type valueResp struct {
+		ID         string `json:"id"`
+		Status     string `json:"status"`
+		Version    string `json:"version"`
+		MinVersion string `json:"min_version"`
+	}
+
+	type response struct {
+		Version  valueResp   `json:"version"`
+		Versions []valueResp `json:"versions"`
+	}
+	var minVersion, maxVersion string
+	var supportedMicroversions SupportedMicroversions
+	var resp response
+	_, err := client.Get(client.Endpoint, &resp, &gophercloud.RequestOpts{
+		OkCodes: []int{200, 300},
+	})
+
+	if err != nil {
+		return supportedMicroversions, err
+	}
+
+	if len(resp.Versions) > 0 {
+		// We are dealing with an unversioned endpoint
+		// We only handle the case when there is exactly one, and assume it is the correct one
+		if len(resp.Versions) > 1 {
+			return supportedMicroversions, fmt.Errorf("unversioned endpoint with multiple alternatives not supported")
+		}
+		minVersion = resp.Versions[0].MinVersion
+		maxVersion = resp.Versions[0].Version
+	} else {
+		minVersion = resp.Version.MinVersion
+		maxVersion = resp.Version.Version
+	}
+
+	// Return early if the endpoint does not support microversions
+	if minVersion == "" && maxVersion == "" {
+		return supportedMicroversions, fmt.Errorf("microversions not supported by ServiceClient Endpoint")
+	}
+
+	supportedMicroversions.MinMajor, supportedMicroversions.MinMinor, err = ParseMicroversion(minVersion)
+	if err != nil {
+		return supportedMicroversions, err
+	}
+
+	supportedMicroversions.MaxMajor, supportedMicroversions.MaxMinor, err = ParseMicroversion(maxVersion)
+	if err != nil {
+		return supportedMicroversions, err
+	}
+
+	return supportedMicroversions, nil
+}
+
+// RequireMicroversion checks that the required microversion is supported and
+// returns a ServiceClient with the microversion set.
+func RequireMicroversion(client gophercloud.ServiceClient, required string) (gophercloud.ServiceClient, error) {
```

Comment:
I like this function signature. You give me a version, and I tell you if it's OK. This requires many less assumptions. Can we make it so that this is the only new exported symbol, and unexport all the rest?

Reply:
Please don't. GopherCloud is already bare bones enough to justify another high-level client built upon it.

Reply:
I may be failing to see your intended use case. Do you have one in mind?

The use case I’m most inclined to think: a client has two max three behaviours coded for a specific action, each corresponding to a specific API. The client would ask for its preferred version, then if not available it would test the fallback behaviour, and so on. In this case the client knows what exact microversion it needs and asks Goohercloud whether Itshak OK. Do you have a case where a client needs something different, like a range?

Reply:
Largely, I want to know *in advance* what features I can and cannot use, so that I don't need to delay this decision until much later in the code. Our code is organized in the way that the controller layer does not know about the Ironic implementation details, and we'd rather keep it that way.